server/src/main/java/org/opensearch/indices/replication/OngoingSegmentReplications.java

```diff
@@ -137,7 +137,7 @@ void startSegmentCopy(GetSegmentFilesRequest request, ActionListener<GetSegmentF
      * @return {@link CopyState} the built CopyState for this replication event.
      * @throws IOException - When there is an IO error building CopyState.
      */
-    CopyState prepareForReplication(CheckpointInfoRequest request, FileChunkWriter fileChunkWriter) throws IOException {
+    synchronized CopyState prepareForReplication(CheckpointInfoRequest request, FileChunkWriter fileChunkWriter) throws IOException {
```

Comment:
Yeah, this class has become a bit of a mess over time so I think its worth cleaning it up.

- Yep this would eventually become an issue as the amount of shards/node and replica count increases, so would be worthwhile to refactor to not block on the object.

I think we can actually do away with the CopyState map at this point altogether and tie an instance & its lifecycle directly to the handler.  The original intention of holding that map was to reduce the amount of disk reads made while computing file metadata, but this is cached internally in IndexShard now.

Reply:
@andrross I've made a revision here that removes the synchronization & copystate map entirely from OngoingSegmentReplications.  This should simply the flow a bit and rely on the map implementation for concurrency where we add/start/remove by allocation ID.

The CopyState object still exists, but its lifecycle is handled by the handler itself, and clears it whenever it completes its work or is cancelled.